src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/Contract.cs

```diff
@@ -42,32 +75,34 @@ public static void ThrowIfFalse([DoesNotReturnIf(parameterValue: false)] bool co
         /// Throws a non-accessible exception if the provided value is true. This method executes in
         /// all builds.
         /// </summary>
-        public static void ThrowIfTrue([DoesNotReturnIf(parameterValue: true)] bool condition, string? message = null)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void ThrowIfTrue([DoesNotReturnIf(parameterValue: true)] bool condition)
         {
             if (condition)
             {
-                message ??= "Unexpected true";
-                Fail(message);
+                Fail("Unexpected true");
             }
         }
 
-        [DebuggerHidden]
-        [DoesNotReturn]
-        public static void Fail(string message = "Unexpected")
+        /// <summary>
+        /// Throws a non-accessible exception if the provided value is true. This method executes in
+        /// all builds.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void ThrowIfTrue([DoesNotReturnIf(parameterValue: true)] bool condition, string message)
         {
-            throw new InvalidOperationException(message);
+            if (condition)
+            {
+                Fail(message);
+            }
         }
 
         [DebuggerHidden]
         [DoesNotReturn]
-        public static T FailWithReturn<T>(string message = "Unexpected")
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
Why the NoInlining attribute here? I mean sure I get that inlining isn't a perf win for this method, but I'm surprised we'd write anything at all.

Reply:
Inlining is not free - it makes NGEN'd code larger. So we don't want to inline code that's only gonna be invoked in exceptional case. 
This is pattern used in CoreFX. See e.g. https://source.dot.net/#System.Reflection.Metadata/System/Reflection/Throw.cs,39c1bafaf83923bf

Reply:
I'll add a comment.